File: src/FarmEngine.Schemas/Quests.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace FarmEngine.Schemas;

// Port of packages/engine-schemas/src/quests.ts.

public sealed record QuestObjective
{
    public string Id { get; init; } = "";
    /// <summary>One of <see cref="QuestObjectiveTypes"/>.</summary>
    public string Type { get; init; } = "";
    public string Description { get; init; } = "";
    public string? TargetItemId { get; init; }
    public double? TargetItemQuantity { get; init; }
    public string? TargetCropType { get; init; }
    public double? TargetCropQuantity { get; init; }
    [JsonPropertyName("targetNPCId")]
    public string? TargetNpcId { get; init; }
    public string? TargetSceneId { get; init; }
    public bool Completed { get; init; }
    public double Progress { get; init; }

    [JsonExtensionData]
    public Dictionary<string, JsonElement>? Extra { get; init; }
}

/// <summary>TS <c>QuestRewardsSchema.items</c> item (inline object).</summary>
public sealed record QuestRewardItem
{
    public string ItemId { get; init; } = "";
    public double Quantity { get; init; }
}

public sealed record QuestRewards
{
    public double? Money { get; init; }
    public List<QuestRewardItem>? Items { get; init; }
    public double? Experience { get; init; }

    [JsonExtensionData]
    public Dictionary<string, JsonElement>? Extra { get; init; }
}

public sealed record Quest
{
    public string Id { get; init; } = "";
    public string Name { get; init; } = "";
    public string Description { get; init; } = "";
    public string? Giver { get; init; }
    /// <summary>One of <see cref="QuestStatuses"/>.</summary>
    public string Status { get; init; } = "";
    public List<QuestObjective> Objectives { get; init; } = [];
    public QuestRewards Rewards { get; init; } = new();
    public List<string>? Prerequisites { get; init; }
    public bool? AutoStart { get; init; }
    public bool? Repeatable { get; init; }
    /// <summary>Seasonal availability (M3): quest only offered/auto-started in these seasons.</summary>
    public List<string>? AvailableSeasons { get; init; }
    /// <summary>Absolute-day window (M3): offered from/until these days (inclusive).</summary>
    public double? AvailableFromDay { get; init; }
    public double? AvailableToDay { get; init; }

    [JsonExtensionData]
    public Dictionary<string, JsonElement>? Extra { get; init; }
}

File: crates/farm-sim/src/schema/quests.rs
//! Port of `Quests.cs` (packages/engine-schemas/src/quests.ts).

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct QuestObjective {
    pub id: String,
    /// One of [`super::quest_objective_types`].
    pub r#type: String,
    pub description: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_item_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_item_quantity: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_crop_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_crop_quantity: Option<f64>,
    #[serde(rename = "targetNPCId", skip_serializing_if = "Option::is_none")]
    pub target_npc_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_scene_id: Option<String>,
    pub completed: bool,
    pub progress: f64,
    #[serde(flatten)]
    pub extra: Map<String, Value>,
}

/// TS `QuestRewardsSchema.items` item (inline object).
#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct QuestRewardItem {
    pub item_id: String,
    pub quantity: f64,
}

#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct QuestRewards {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub money: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub items: Option<Vec<QuestRewardItem>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub experience: Option<f64>,
    #[serde(flatten)]
    pub extra: Map<String, Value>,
}

#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct Quest {
    pub id: String,
    pub name: String,
    pub description: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub giver: Option<String>,
    /// One of [`super::quest_statuses`].
    pub status: String,
    pub objectives: Vec<QuestObjective>,
    pub rewards: QuestRewards,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prerequisites: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auto_start: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub repeatable: Option<bool>,
    /// Seasonal availability (M3): quest only offered/auto-started in these seasons.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub available_seasons: Option<Vec<String>>,
    /// Absolute-day window (M3): offered from/until these days (inclusive).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub available_from_day: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub available_to_day: Option<f64>,
    #[serde(flatten)]
    pub extra: Map<String, Value>,
}
